kernel/src/fs/cgroupfs/controller/mod.rs

```diff
@@ -0,0 +1,375 @@
+            return Err(Error::IsDead);
+        };
+
+        controller.read_attr_at(name, offset, writer)
+    }
+}
+
+/// A locked controller for a cgroup.
+///
+/// Holding this lock indicates exclusive access to modify the sub-control state.
+pub(super) struct LockedController<'a> {
+    active_set: MutexGuard<'a, SubCtrlSet>,
+    controller: &'a Controller,
+}
+
+impl LockedController<'_> {
+    /// Activates a sub-control of the specified type.
+    pub(super) fn activate(
+        &mut self,
+        ctrl_type: SubCtrlType,
+        current_node: &dyn CgroupSysNode,
+        parent_controller: Option<&LockedController>,
+    ) -> Result<()> {
+        if self.active_set.contains_type(ctrl_type) {
+            return Ok(());
+        }
+
+        // A cgroup can activate the sub-control only if this
+        // sub-control has been activated in its parent cgroup.
+        if parent_controller
+            .is_some_and(|controller| !controller.active_set.contains_type(ctrl_type))
+        {
+            return Err(Error::NotFound);
+        }
+
+        self.active_set.add_type(ctrl_type);
+        self.update_sub_controllers_for_children(ctrl_type, current_node);
+
+        Ok(())
+    }
+
+    /// Deactivates a sub-control of the specified type.
+    pub(super) fn deactivate(
+        &mut self,
+        ctrl_type: SubCtrlType,
+        current_node: &dyn CgroupSysNode,
+    ) -> Result<()> {
+        if !self.active_set.contains_type(ctrl_type) {
+            return Ok(());
+        }
+
+        // If any child node has activated this sub-control,
+        // the deactivation operation will be rejected.
+        let mut can_deactivate = true;
+        // This is race-free because if a child wants to activate a sub-controller, it should first
+        // acquire the lock of the parent controller, which is held here.
+
+        for child in current_node.children() {
+            let cgroup_child = child.as_any().downcast_ref::<CgroupNode>().unwrap();
+            let child_controller = cgroup_child.controller().lock();
+            if child_controller.active_set().contains_type(ctrl_type) {
+                can_deactivate = false;
+                break;
+            }
+        }
+        if !can_deactivate {
+            return Err(Error::InvalidOperation);
+        }
+
+        self.active_set.remove_type(ctrl_type);
+        self.update_sub_controllers_for_children(ctrl_type, current_node);
+
+        Ok(())
+    }
+
+    fn update_sub_controllers_for_children(
+        &self,
+        ctrl_type: SubCtrlType,
+        current_node: &dyn CgroupSysNode,
+    ) {
+        current_node.visit_children_with(0, &mut |node| {
+            let cgroup_node = node.as_any().downcast_ref::<CgroupNode>().unwrap();
+            match ctrl_type {
+                SubCtrlType::Memory => {
+                    let new_controller = SubController::new(Some(self));
+                    cgroup_node.controller().memory.update(new_controller);
```

Comment:
We need to update all the descents to make sure the `parent` field is valid.

But it seems that we don't need to have a global lock here.
 1. When we activate or deactivate the controller, we ensure that none of the descents' controllers are activated and cannot be activated concurrently. Therefore, we don't need to worry about concurrent controller activation or deactivation.
 2. Note that we do need to worry about concurrent child addition or deletion. However, as long as we update the controller before checking the children, it is fine, even if we miss new children. New children are guaranteed to see the new controller (synchronized by the child lock).

Is it correct? This needs more explanation in the code comments.